libraries/stdlib/src/kotlin/text/Strings.kt

```diff
@@ -1162,3 +1162,97 @@ public fun CharSequence.lineSequence(): Sequence<String> = splitToSequence("\r\n
  */
 public fun CharSequence.lines(): List<String> = lineSequence().toList()
+
+/**
+ * For each i in [0, length), this function computes
+ * the length of the longest suffix of a substring of pattern from 0 to i
+ * that is also a prefix of the pattern itself.
+ */
+private fun computePrefixFunction(pattern: CharSequence): IntArray {
+    val resultTable = IntArray(pattern.length)
+
+    var matches = 0
+    for (i in 1..pattern.length - 1) {
+        while (matches > 0 && pattern[matches] != pattern[i]) {
+            matches = resultTable[matches - 1]
+        }
+
+        if (pattern[matches] == pattern[i]) {
+            matches++
+        }
+        resultTable[i] = matches
+    }
+
+    return resultTable
+}
+
+/**
+ * Returns a sequence of indices where the char occurs in this CharSequence.
+ *
+ * @param [char] The character to look for in the string
+ * @param [ignoreCase] If true, characters are matched even if one is upper and the other is
+ *                 lower case
+ *
+ * @return A list of indices where the supplied [char] occurs in the text.
+ */
+public fun CharSequence.occurrencesOf(char: Char, ignoreCase: Boolean = false): Sequence<Int> {
+    return indices.asSequence().filter { this[it].equals(char, ignoreCase) }
+}
+
+/**
+ * Returns a sequence of indices where the pattern occurs in this CharSequence. This method
+ * searches character by character and thus does not support regular expressions
+ * as input for the pattern.
+ * For a pattern longer than the text, an empty sequence is returned. If the pattern
+ * is empty, all indices plus an empty occurrence at the end are matched.
+ *
+ * @param [pattern] The pattern to look for in this String. Regular expressions
+ *                 are not supported
+ * @param [ignoreCase] If true, characters are matched even if one is upper and the other is
+ *                 lower case
+ * @param [matchOverlapping] If true, also match overlapping occurrences.
+ *
+ * @return A list of indices where the supplied [pattern] starts in the text.
+ */
+public fun CharSequence.occurrencesOf(pattern: CharSequence, ignoreCase: Boolean = false, matchOverlapping: Boolean = false): Sequence<Int> {
+
+    if (pattern.length > this.length) {
+        return emptySequence()
+    }
+
+    if (pattern.isEmpty()) {
+        return (0..length).asSequence()
+    }
+
+    if (pattern.length == 1) {
+        return occurrencesOf(pattern[0], ignoreCase)
+    }
+
+    // Non-trivial pattern matching, perform computation
+    // using Knuth-Morris-Pratt
+
+    val prefixFunction = computePrefixFunction(pattern)
+
+    var i = 0
+    var matches = 0
+    return generateSequence {
+        while (i < length) {
+            while (matches > 0 && !pattern[matches].equals(this[i], ignoreCase)) {
+                matches = prefixFunction[matches - 1]
+            }
+
+            if (pattern[matches].equals(this[i], ignoreCase)) {
+                matches++
+            }
```

Comment:
This approach requires two character comparisons per position when `matches > 0`. Could you restructure the loop body? Also applies to `computePrefixFunction`.